quic/congestion_control/CongestionControlLocalEnv.cpp

```diff
@@ -93,7 +90,16 @@ void CongestionControlLocalEnv::loop() {
     core_state = outputs->elements()[1].toTuple();
 
     action.cwndAction = *action_tensor.data_ptr<long>();
-    onAction(action);
+
+    // If there is an ongoing shutdown, it is important not to trigger the action
+    // because `onAction()` calls `runImmediatelyOrRunInEventBaseThreadAndWait()`
+    // and this method will hang forever during shutdown, preventing the thread from
+    // exiting cleanly.
+    if (!shutdown_) {
+      onAction(action);
+    } else {
+      LOG(INFO) << "Skipping action due to shutdown in progress";
```

Comment:
Also just return from the loop here?

Reply:
We could -- but looking at it more closely, I realized that in both situations we would "lose" this log:
```cpp
  LOG(INFO) << "Inference loop terminating after " << episode_step
            << " steps, total return = " << episode_return;
```
==> instead I prefer https://github.com/odelalleau/mvfst/pull/2/files where regardless of when the shutdown is detected, we exit the loop then print that log (and not having a `break;` statement after `"Skipping action due to shutdown in progress"` is intended so that `episode_step` is properly incremented (since we already increased `episode_return`).